bigquery_etl/docs/mozfun/generate_mozfun_docs.py

```diff
@@ -41,6 +44,28 @@ def add_source_and_edit(source_url, edit_url):
     return f"[Source]({source_url})  |  [Edit]({edit_url})"
 
 
+def get_inputs_outputs_from_udf(root: str) -> Tuple[List[str], List[str]]:
+    """Parse UDF SQL file and return inputs and outputs lists."""
+    with open(os.path.join(root, UDF_FILE), "r") as udf_file:
+        input_str, output_str = get_udf_parameters(udf_file.read())
+
+    input_lines = []
+    if input_str is not None:
+        if "<" in input_str and ">" in input_str:
```

Comment:
Since the function get_udf_parameters already returns the formatted parameters, perhaps we can avoid splitting and concatenating again by adding this validation and a test to that function?

Reply:
So I agree that would simplify things here, but I think it depends on how we think the utility function might be used. I thought that doing the formatting here helped keep the utility more generic, but another issue is that I also couldn't immediately think of a solution to the various splitting issues I came across. I'm sure we could update the regex to pull out ARRAY/STRUCT/etc types, but I was trying to balance the time I'd need to spend on that with the priority of the issue. This solution seemed quickest to getting an acceptable (to me, at least) result. Another alternative would be to simply never split out inputs by line, but having one input per line seems nicer to me.

What do you think?

Reply:
I think I would opt for just using the output you get for now. Because having it sometimes in different lines and sometimes as one line might be confusing / inconsistent. 

If we get bored at some point we could try to tackle the parsing of ARRAY/STRUCT values .. it might be not regex but stacks? somehow every coding puzzle with parentheses always comes back to stacks 

But since I'm not sure how it should look like in the end, I can't think of how to parse it yet. 

So my take would be: let's move the formatting to the utils and have it return a string (with or without line breaks) and if we ever end up implementing it we wouldn't need to touch the `generate_mozfun_docs.py` again? 

I don't think the utils functions are generic enough to be used for anything else. It's very specific regex for exactly that purpose?

Reply:
Ok, I will update to remove some of the formatting I'm doing now for consistency's sake. I think if the functions are not intended to be generic they may be more appropriately placed either in the `generate_mozfun_docs.py` file or alongside it.